core/src/main/scala/kafka/server/ClientMetricsManager.scala

```diff
@@ -62,11 +108,92 @@ class ClientMetricsManager {
 
     new GetTelemetrySubscriptionResponse(data)
   }
 
+  def validatePushRequest(pushTelemetryRequest: PushTelemetryRequest,
+                          clientInfo: CmClientInformation): Unit = {
+
+    def isSupportedCompressionType(id: Int) : Boolean = {
+      try {
+        CompressionType.forId(id)
+        true
+      } catch {
+        case e: IllegalArgumentException => false
+      }
+    }
+
+    val clientInstanceId = Option(pushTelemetryRequest.getClientInstanceId) match {
+      case None | Some(Uuid.ZERO_UUID) =>
+        val msg = String.format("Invalid request from the client [%s], missing client instance id", clientInfo.getClientId)
+        throw new ClientMetricsException(msg, Errors.INVALID_REQUEST)
+      case _ => pushTelemetryRequest.getClientInstanceId
+    }
+
+    val clientInstance = getClientInstance(clientInstanceId).getOrElse(createClientInstance(clientInstanceId, clientInfo))
+    if (pushTelemetryRequest.getSubscriptionId != clientInstance.getSubscriptionId) {
+      val msg = String.format("Client's subscription id [%s] != Broker's cached client's subscription id [%s]",
+        pushTelemetryRequest.getSubscriptionId.toString, clientInstance.getSubscriptionId.toString)
+      throw new ClientMetricsException(msg, Errors.UNKNOWN_CLIENT_METRICS_SUBSCRIPTION_ID)
+    }
+
+    if (!isSupportedCompressionType(pushTelemetryRequest.data().compressionType)) {
+      val msg = String.format("Unknown compression type [%s] is received in PushTelemetryRequest from %s",
+        pushTelemetryRequest.data().compressionType().toString, pushTelemetryRequest.getClientInstanceId.toString)
+      throw new ClientMetricsException(msg, Errors.UNSUPPORTED_COMPRESSION_TYPE)
+    }
+
+    if (!clientInstance.canAcceptPushRequest(pushTelemetryRequest.isClientTerminating)) {
+      val msg = String.format("Request from the client [%s] arrived before the throttling time",
+        pushTelemetryRequest.getClientInstanceId.toString)
+      throw new ClientMetricsException(msg, Errors.THROTTLING_QUOTA_EXCEEDED)
+    }
+  }
+
+  def processPushTelemetryRequest(pushTelemetryRequest: PushTelemetryRequest,
+                                  requestContext: RequestContext,
+                                  clientInfo: CmClientInformation,
+                                  throttleMs: Int): PushTelemetryResponse = {
+
+    def createResponse(errors: Option[Errors]) : PushTelemetryResponse = {
+      var adjustedThrottleMs = throttleMs
+      val clientInstance = getClientInstance(pushTelemetryRequest.getClientInstanceId)
+
+      // Before sending the response make sure to update the book keeping markers like
+      // lastAccessTime, isTerminating flag etc..
+      if (!clientInstance.isEmpty) {
+        adjustedThrottleMs = Math.max(clientInstance.get.getAdjustedPushInterval(), throttleMs)
+        clientInstance.get.updateLastAccessTs(getCurrentTime)
+        clientInstance.get.setTerminatingFlag(pushTelemetryRequest.isClientTerminating)
```

Comment:
What if a client does:
 1. Push (terminating=true)
 2. Push(terminating=false)

I think we should perhaps return an error in this case since it is a breach of the protocol, what do you think?

Reply:
Agreed, we should return error

Reply:
I made it simple, once client sets the state to Terminating only once request is accepted any further requests would be rejected.